README.md
# bkash-payment-api

Nodejs library to accept bkash payments on your backend application. BKASH API WRAPPER

[![CI](https://github.com/siyamhosan/bkash-payment-api/actions/workflows/main.yml/badge.svg)](https://github.com/siyamhosan/bkash-payment-api/actions/workflows/main.yml)
[![Publish](https://github.com/siyamhosan/bkash-payment-api/actions/workflows/publish.yml/badge.svg)](https://github.com/siyamhosan/bkash-payment-api/actions/workflows/publish.yml)

## Examples

- [Express](https://github.com/shahriar-shojib/bkash-payment-gateway/tree/main/examples) from main Author [Shahriar Shojib](https://github.com/shahriar-shojib) *may not work*

## Features

- Implements all the methods required to get accepted as a merchant on bKash
- Written in typescript
- Get intellisense in when interacting with the library `vscode`
- Get Documentation and examples right inside your code editor `vscode`
- Get Bkash Response Intellisense
- Abort Request when an executePayment request exceeds `30 seconds` and get an error so that you can query the payment
- Get Human Readable exceptions when some error response is returned from bKash `in progress`

---

# How to use

## Installing the library

### `npm`

> `npm install bkash-payment-api`

### `yarn`

> `yarn add bkash-payment-api`

---

## Getting Bkash Credentials
****You need Bkash Merchant or Retails account to use this library**
1. Goto [Bkash PGW](https://pgw-integration.bkash.com) and login or create a new account
2. [Request for New Product](https://pgw-integration.bkash.com/#/merchant/new-product-request)
3. Fill up the form and submit
 > **Select Application Platform**: select all ✔
4. Get [Sandbox Credentials](https://pgw-integration.bkash.com/#/merchant/sandbox-credentials) these are your bkash credentials/config
> baseUrl is `https://checkout.sandbox.bka.sh/v1.2.0-beta` for sandbox
> baseUrl is `https://tokenized.pay.bka.sh/v1.2.0-beta` for production
5. [Validate Sandbox](https://pgw-integration.bkash.com/#/merchant/sandbox-validation) Credentials for Live Production Credentials 
> 1. [Create a payment](#create-a-payment)
> 2. get the response and enter that in `Create Payment Sandbox Test` also get the `paymentID` from response
> 3. [Execute a payment](#execute-a-payment-with-payment-id) with the `paymentID` and enter the response in `Execute Payment Sandbox Test`
> 4. Submit and You got your [Live Credentials](https://pgw-integration.bkash.com/#/merchant/live-credentials)

## Initializing the library

### `javascript`

> file `bkash.js`

```javascript
const { BkashGateway } = require('bkash-payment-api');

const bkashConfig = {
	baseURL: 'https://checkout.sandbox.bka.sh/v1.2.0-beta', //do not add a trailing slash
	key: 'abcdxx2369',
	username: 'bkashTest',
	password: 'bkashPassword1',
	secret: 'bkashSup3rS3cRet',
};

const bkash = new BkashGateway(config);
module.exports = bkash;
```

### `typescript`

> file `bkash.ts`

```typescript
import { BkashGateway IBkashConstructor } from 'bkash-payment-api';

const bkashConfig: IBkashConstructor = {
	//get intellisense here
	baseURL: 'https://checkout.sandbox.bka.sh/v1.2.0-beta', //do not add a trailing slash
	key: 'abcdxx2369',
	username: 'bkashTest',
	password: 'bkashPassword1',
	secret: 'bkashSup3rS3cRet',
};

const bkash = new BkashGateway(config);
export default bkash;
```

---

## Create a payment

```javascript
const paymentRequest = {
	amount: 1000,
	orderID: 'ORD1020069',
	intent: 'sale',
	callBack: 'https://example.com/bkash/callback',
};

const result = await bkash.createPayment(paymentRequest);
console.log(result);
```

---

## Execute a payment with payment ID

```javascript
const paymentId = createPaymentRequest.paymentID; // sample payment ID
const result = await bkash.executePayment(paymentId);
```

---

## Query a payment with payment ID

```javascript
const paymentId = createPaymentRequest.paymentID; // sample payment ID
const result = await bkash.queryPayment(paymentId);
```

---

## Search Transaction

```javascript
const result = await bkash.searchTransaction('TRX22347463XX');
```

---

## Refund a transaction
Not Working Right Now

```javascript
const refundTransactionData = {
	paymentID: '22423169',
	amount: '25.69', //do not add more than two decimal points
 	trxID: 'TRX22347463XX';
 	sku: 'SK256519';
}

const result = await bkash.refundTransaction(refundTransactionData);
```

---

## Check Refund Status
Not Working Right Now

```javascript
const result = await bkash.refundStatus('TRX22347463XX', '12437969');
```

---

### Contributing

- Please Follow the code style and use the prettier config and eslint config provided in the repository
- Feel free to open `issues` or `pull request` for any issues and bugfixes
- If you want to implement new features or write documentation about existing features feel free to do it as well
- To see a list of missing features or to-do's, please visit the `project` section of the github repository

---

### License

> MIT

> DISCLAIMER: This software comes with absolutely no warranty and is not affiliated with the company **`Bkash`** in any way. Use at your own risk. Author and Contributors are not responsible for any financial damages, outages etc.

### Author
Continue By
[Siyam Hosan](https://github.com/siyamhosan)
Created by
[Shahriar Shojib](https://github.com/shahriar-shojib)

CHANGELOG.md
# bkash-payment-api

## 1.0.1

### Patch Changes

- fbd0b1f: Fixes

## 1.0.0

### Major Changes

- 7688abf: Sable version

### Patch Changes

- 9ca4d7c: Bug Fixed and Stable

## 0.9.2

### Patch Changes

- 3d026da: Imports And Exports

## 0.9.1

### Patch Changes

- dc28165: tested

## null

### Patch Changes

- 23b913e: v1

## 0.0.3

### Patch Changes

- 845a4c7: new sample package

## 0.0.2

### Patch Changes

- 3b5d3a8: added a sample function
- 8b596b8: Intial


src/tests/bkash.spec.ts
import { config } from 'dotenv';
import BkashGateway from '..';
import { BkashException } from '../exceptions/bkashException';

config();

jest.setTimeout(30000);

describe('Bkash API', () => {
	const bkash = new BkashGateway({
		baseURL: process.env.BKASH_BASEURL!,
		key: process.env.BKASH_API_KEY!,
		secret: process.env.BKASH_API_SECRET!,
		password: process.env.BKASH_PASSWORD!,
		username: process.env.BKASH_USERNAME!,
	});

	it('should create a payment', async () => {
		const payment = await bkash.createPayment({
			amount: 100,
			intent: 'sale',
			orderID: 'ORD1020069',
		});

		expect(payment).toBeDefined();
		expect(payment.amount).toBe('100');
		expect(payment.intent).toBe('sale');
		expect(payment.paymentID).toBeDefined();
	});

	it('should be able to query a payment by paymentID', async () => {
		const payment = await bkash.createPayment({
			amount: 100,
			intent: 'sale',
			orderID: 'ORD1020069',
		});

		const paymentQuery = await bkash.queryPayment(payment.paymentID);

		expect(paymentQuery).toBeDefined();

		expect(paymentQuery.amount).toBe('100');
		expect(paymentQuery.intent).toBe('sale');

		expect(paymentQuery.transactionStatus).toBe('Initiated');
		expect(paymentQuery.merchantInvoiceNumber).toBe('ORD1020069');
	});

	it('should fail to execute payment if invalid payment ID was provided', async () => {
		let success = false;
		try {
			await bkash.executePayment('PAYMENT_ID');
			success = true;
		} catch (error) {
			expect(error).toStrictEqual(new BkashException('Invalid Payment ID'));
		}
		expect(success).toBe(false);
	});

	it('should be fail to execute payment if the user did not authorize the payment', async () => {
		const payment = await bkash.createPayment({
			amount: 100,
			intent: 'sale',
			orderID: 'ORD1020069',
		});

		let success = false;
		try {
			await bkash.executePayment(payment.paymentID);
			success = true;
		} catch (error) {
			expect(error).toStrictEqual(new BkashException('Invalid Payment ID'));
		}
		expect(success).toBe(false);
	});
});


src/index.ts
import { BkashGateway } from './bkashGateway'
import { BkashException } from './exceptions/bkashException'
import {
  IBkashCreatePaymentResponse,
  IBkashExecutePaymentResponse,
  IBkashQueryPaymentResponse,
  IBkashRefundResponse,
  IBkashRefundStatusResponse,
  IBkashSearchTransactionResponse,
  IBkashTokenResponse
} from './interfaces/bkashResponse.interface'
import { ICreatePayment } from './interfaces/createPayment.interface'
import { IHeaders } from './interfaces/headers.interface'
import { IBkashConstructor, IRefundArgs } from './interfaces/main.interface'

export {
  BkashGateway,
  BkashException,
  IBkashCreatePaymentResponse,
  IBkashExecutePaymentResponse,
  IBkashQueryPaymentResponse,
  IBkashRefundResponse,
  IBkashRefundStatusResponse,
  IBkashSearchTransactionResponse,
  IBkashTokenResponse,
  ICreatePayment,
  IHeaders,
  IBkashConstructor,
  IRefundArgs
}
export default BkashGateway


src/interfaces/bkashResponse.interface.ts
import { Intent } from './createPayment.interface';

export interface IBkashTokenResponse {
	expires_in: string;
	id_token: string;
	refresh_token: string;
	token_type: string;
	status?: string;
	msg?: string;
}

export interface IBkashCreatePaymentResponse {
	paymentID: string;
	createTime: string;
	orgLogo: string;
	orgName: string;
	transactionStatus: string;
	amount: string;
	currency: string;
	intent: string;
	merchantInvoiceNumber: string;
	//need to implement error cases here
}

export interface IBkashExecutePaymentResponse {
	statusCode: '0000' | string;
	statusMessage: 'Successful' | string;
	paymentID: string;
	customerMsisdn: string;
	payerReference: string;
	paymentExecuteTime: string;
	trxID: string;
	transactionStatus: 'Completed' | string;
	amount: string;
	currency: 'BDT' | string;
	intent: 'sale' | string;
	merchantInvoiceNumber: string;
}

export interface IBkashQueryPaymentResponse {
	paymentId: string;
	createTime: string;
	updateTime: string;
	trxID: string;
	transactionStatus: string;
	amount: string;
	currency: string;
	intent: Intent;
	merchantInvoiceNumber: string;
	refundAmount: string;
}

export interface IBkashSearchTransactionResponse {
	trxID: string;
	transactionStatus: string;
	transactionType: string;
	amount: string;
	currency: string;
	customerMsisdn: string;
	organizationShortCode: string;
	initiationTime: string;
	completedTime: string;
}

export interface IBkashRefundResponse {
	completedTime: string;
	transactionStatus: string;
	originalTrxID: string;
	refundTrxID: string;
	amount: string;
	currency: string;
	charge: string;
}

export interface IBkashRefundStatusResponse {
	completedTime: string;
	transactionStatus: string;
	originalTrxID: string;
	refundTrxID: string;
	amount: string;
	currency: string;
	charge: string;
}


src/bkashGateway.ts
import { IHeaders } from './interfaces/headers.interface'
import { diffSeconds } from './utils/diffSeconds'
import { get, post } from './utils/request'

import {
  IBkashCreatePaymentResponse,
  IBkashExecutePaymentResponse,
  IBkashQueryPaymentResponse,
  IBkashRefundResponse,
  IBkashRefundStatusResponse,
  IBkashSearchTransactionResponse,
  IBkashTokenResponse
} from './interfaces/bkashResponse.interface'

import { BkashException } from './exceptions/bkashException'
import { ICreatePayment } from './interfaces/createPayment.interface'
import { IBkashConstructor, IRefundArgs } from './interfaces/main.interface'

/**
 * Bkash Payment Gateway Main Entrypoint
 * @example
 * ```javascript
 * const BkashGateway = require('bkash-payment-gateway');
 * const bkash = new BkashGateway({
 *		baseURL: process.env.BKASH_BASEURL,
 *		key: process.env.BKASH_API_KEY,
 *		secret: process.env.BKASH_API_SECRET,
 *		username: process.env.BKASH_USERNAME,
 *		password: process.env.BKASH_PASSWORD,
 * });
 * ```
 */
export class BkashGateway {
  private token!: string
  private refreshToken!: string
  private tokenIssueTime!: number
  private readonly secret: string
  private readonly key: string
  private readonly baseURL: string
  private headers: IHeaders
  /**
   *
   * @param config config object required by the `bkash-payment-gateway` package
   * @example
   * ```
   * const bkashConfig = {
   *   baseURL: 'https://tokenized.sandbox.bka.sh/v1.2.0-beta', //sandbox
   *   key: 'abcdxx2369',
   *   username: 'bkashTest',
   *   password: 'bkashPassword1',
   *   secret: 'bkashSup3rS3cRet',
   * }
   * const bkash = new BkashGateway(config)
   * ```
   *
   */

  constructor (config: IBkashConstructor) {
    // validations
    if (Object.keys(config).length !== 5)
      throw new BkashException('Invalid Configuration provided')
    this.validateConfig(config)

    //main task
    const { baseURL, key, password, secret, username } = config
    this.baseURL = baseURL
    this.key = key
    this.secret = secret
    this.headers = {
      username,
      password
    }
  }

  /**
   * Start the initial payment request
   *
   * @param paymentDetails Information required to start a payment flow
   *
   * @returns Promise of Bkash Create payment Response
   * @example
   * ```
   * const result = await bkash.createPayment({
   *   amount: 1000,
   *   orderID: 'ORD1020069',
   *   intent: 'sale',
   *   callBack: 'https://example.com/bkash/callback',
   * });
   * ```
   */
  public createPayment = async (
    paymentDetails: ICreatePayment
  ): Promise<IBkashCreatePaymentResponse> => {
    const { amount, intent, orderID, callBack } = paymentDetails

    const payload = {
      orderId: orderID,
      amount,
      intent,
      mode: '0011',
      currency: 'BDT',
      merchantInvoiceNumber: orderID,
      payerReference: orderID,
      callbackURL: callBack
    }

    const headers = await this.createTokenHeader()
    return await post<IBkashCreatePaymentResponse>(
      `${this.baseURL}/tokenized/checkout/create`,
      payload,
      headers
    )
  }

  /**
   * Execute a payment after a user has completed bkash auth flow
   * @param paymentID - Payment ID to Execute
   * @example
   * ```
   * const result = await bkash.executePayment(paymentID);
   * ```
   */
  public executePayment = async (
    paymentID: string
  ): Promise<IBkashExecutePaymentResponse> => {
    try {
      const headers = await this.createExecuteHeader()
      return await post<IBkashExecutePaymentResponse>(
        `${this.baseURL}/tokenized/checkout/execute`,
        { paymentID },
        headers
      )
    } catch (error) {
      if (error instanceof BkashException) {
        throw error
      }

      throw new BkashException(
        'Timeout of 30 Seconds Exceeded While Executing Payment, Please Query the Payment'
      )
    }
  }

  /**
   * Query Payment From Bkash
   * @param paymentID - Payment ID to Query
   *
   * @example
   * ```
   * const result = await bkash.queryPayment(paymentID);
   * ```
   */
  public queryPayment = async (
    paymentID: string
  ): Promise<IBkashQueryPaymentResponse> => {
    const headers = await this.createTokenHeader()
    return await get<IBkashQueryPaymentResponse>(
      `${this.baseURL}/tokenized/checkout/payment/status`,
      headers,
      {
        paymentID
      }
    )
  }

  /**
   * Search with a transaction ID
   * @param trxID - Transaction ID to Search
   *
   * @example
   * ```
   * const result = await bkash.searchTransaction('TRX22347463XX');
   * ```
   */
  public searchTransaction = async (
    trxID: string
  ): Promise<IBkashSearchTransactionResponse> => {
    return await get<IBkashSearchTransactionResponse>(
      `${this.baseURL}/tokenized/checkout/general/searchTran`,
      await this.createTokenHeader(),
      { trxID }
    )
  }

  /**
   * Refund a transaction
   * @param trxID - Transaction ID to Search
   * @deprecated
   * @example
   * ```
   * const refunTransactionData = {
   *  paymentID: '22423169',
   *  amount: '25.69', //do not add more than two decimal points
   *  trxID: 'TRX22347463XX',
   *  sku: 'SK256519',
   * };
   *
   * const result = await bkash.refundTransaction(refunTransactionData);
   * ```
   */
  public refundTransaction = async (
    refundInfo: IRefundArgs
  ): Promise<IBkashRefundResponse> => {
    return post<IBkashRefundResponse>(
      `${this.baseURL}/tokenized/checkout/payment/refund`,
      refundInfo,
      await this.createTokenHeader()
    )
  }

  /**
   * Check Refund Status for a given paymentID and transaction ID
   * @param trxID transaction ID
   * @param paymentID payment ID
   * @deprecated
   * @example
   * ```
   * const result = await bkash.refundStatus('TRX22347463XX', '12437969');
   * ```
   */

  public refundStatus = async (
    trxID: string,
    paymentID: string
  ): Promise<IBkashRefundStatusResponse> => {
    return await post<IBkashRefundStatusResponse>(
      `${this.baseURL}/tokenized/checkout/payment/refund`,
      { trxID, paymentID },
      await this.createTokenHeader()
    )
  }

  private createTokenHeader = async (isBearer?: boolean): Promise<IHeaders> => {
    const token = await this.getToken()
    isBearer = isBearer ?? true
    return {
      authorization: isBearer ? `Bearer ${token}` : token,
      'x-app-key': this.key,
      accept: 'application/json',
      'content-type': 'application/json'
    }
  }
  private createExecuteHeader = async () => {
    const token = await this.getToken()
    return {
      Authorization: token,
      'x-app-key': this.key,
      accept: 'application/json',
      'content-type': 'application/json'
    }
  }
  private getToken = async (): Promise<string> => {
    if (!this.token) {
      const { id_token, refresh_token, msg, status } =
        await this.getInitialToken()

      //throw error if bkash sends status [only happens when request fails]
      if (status && msg) throw new BkashException(msg)

      this.token = id_token
      this.refreshToken = refresh_token
      this.tokenIssueTime = Date.now()
      return this.token
    }

    const diff = diffSeconds(this.tokenIssueTime)

    if (diff < 3500) {
      return this.token
    }

    //token is expired, refresh it
    const { id_token, refresh_token, msg, status } = await this.newToken(
      this.refreshToken
    )

    //throw error if bkash sends status [only happens when request fails]
    if (status && msg) throw new BkashException(msg)

    this.token = id_token
    this.refreshToken = refresh_token
    this.tokenIssueTime = Date.now()
    return this.token
  }

  private getInitialToken = async (): Promise<IBkashTokenResponse> => {
    const response = await post<IBkashTokenResponse>(
      `${this.baseURL}/tokenized/checkout/token/grant`,
      {
        app_key: this.key,
        app_secret: this.secret
      },
      this.headers
    )
    if (response.status === 'fail')
      throw new BkashException('Invalid API Credentials Provided')
    return response
  }

  private newToken = (refresh: string): Promise<IBkashTokenResponse> => {
    return post<IBkashTokenResponse>(
      `${this.baseURL}/tokenized/checkout/token/refresh`,
      {
        app_key: this.key,
        app_secret: this.secret,
        refresh_token: refresh
      },
      this.headers
    )
  }

  private validateConfig = (config: IBkashConstructor): void => {
    const { baseURL, key, password, secret, username } = config

    if (!baseURL || baseURL === '')
      throw new BkashException('Invalid BaseURL provided')
    if (!key || key === '') throw new BkashException('Invalid API Key provided')
    if (!secret || secret === '')
      throw new BkashException('Invalid API secret provided')
    if (!username || username === '')
      throw new BkashException('Invalid API username provided')
    if (!password || password === '')
      throw new BkashException('Invalid API password provided')
  }
}
